### PR TITLE
add stream-k v0.2

### PR DESCRIPTION
**Pull request by xiaohuguo2023**
streamk v0.2:

- new streamk tuning script to reduce compiling and profiling time

- use load/store cache modifier to reimplement spinning lock

- add CI test for streamk-kernel

- able to use streampipelineV2

### CODE REVIEW

File: python/perf-kernels/streamk/streamk_kernel.py [hunk 1]
```diff
@@ -166,25 +146,34 @@ def streamk_gemm(
...
-            tile_iter_end = tile_iter + iters_per_tile
+
+        if start_iter != tile_iter:
+            rm1 = tl.arange(0, BLOCK_SIZE_M)
+            rn1 = tl.arange(0, BLOCK_SIZE_N)
+            rm1 = tl.max_contiguous(tl.multiple_of(rm1, BLOCK_SIZE_M), BLOCK_SIZE_M)
+            rn1 = tl.max_contiguous(tl.multiple_of(rn1, BLOCK_SIZE_N), BLOCK_SIZE_N)
+            P_ = P + pid * BLOCK_SIZE_M * BLOCK_SIZE_N + rm1[:, None] * BLOCK_SIZE_N + rn1[None, :]
+            tl.store(P_, acc, cache_modifier=".wt")
+            tl.store(locks + pid, 1, cache_modifier=".wt")
```
Comment (neoblizz) [inline python/perf-kernels/streamk/streamk_kernel.py:160]: Copying some of my notes again on gfx90a: For gfx90a the load/stores with `cache_modifiers` do not work. Documented here: https://github.com/ROCm/triton-internal/issues/311

File: python/perf-kernels/streamk/streamk_kernel.py [hunk 2]
```diff
@@ -166,25 +146,34 @@ def streamk_gemm(
...
+        else:
             next_pid = pid + 1
+            tile_iter_end = tile_iter + iters_per_tile
             end = end_iter
-            while (end < tile_iter_end and next_pid < num_cus):
-                # todo: try use tl.load once cache modifier landed upstream
-                while tl.atomic_cas(locks + next_pid, 1, 1) != 1:
+            while (end < tile_iter_end and next_pid < NUM_SMS):
+                while tl.load(locks + next_pid, cache_modifier=".cv", volatile=True) != 1:
```
Comment (neoblizz) [inline python/perf-kernels/streamk/streamk_kernel.py:166]: This also does not work in gfx90a: https://github.com/ROCm/triton-internal/issues/311

File: python/perf-kernels/streamk/streamk_kernel.py [hunk 3]
```diff
@@ -2,121 +2,100 @@
...
     BLOCK_SIZE_N: tl.constexpr,
     BLOCK_SIZE_K: tl.constexpr,
     GROUP_SIZE_M: tl.constexpr,
+    NUM_SMS: tl.constexpr,
+    BIAS: tl.constexpr,
     EVEN_K: tl.constexpr,
 ):
     pid = tl.program_id(0)
-    pid = get_new_pid(pid, num_cus)
+    pid = (pid % 8) * (NUM_SMS // 8) + (pid // 8)
```
Comment (neoblizz) [inline python/perf-kernels/streamk/streamk_kernel.py]: This is not needed for anything but gfx942, so we will actually remove this if the arch was gfx90a. So we need a guard: if == gfx942 do this, else nothing.

File: python/perf-kernels/streamk/utils/file_generator.py [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+    runs = iters if run_bench else 120
+    for config in configs:
+        configStr = gen_configStr(config)
+        block_m = config.get('BLOCK_SIZE_M')
+        block_n = config.get('BLOCK_SIZE_N')
+        num_sms = config.get('NUM_SMS')
+        matmul_call_str = f"""
+    if '{configStr}' not in failed_configs:
+        rotating_num = tensors['rotating_num']
+        locks = torch.zeros(({runs}, {num_sms}), device = "cuda", dtype = torch.int32)
```
Comment (neoblizz) [inline python/perf-kernels/streamk/utils/file_generator.py:341]: `locks` can be less than `int32` type, we only need 1 byte: `uint8` should work. Tensile uses `uint8`.

File: python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
+import torch
+import triton
+import random
+
+from streamk_kernel import streamk_gemm
+#from persistent_loop import streamk_gemm
```
Comment (neoblizz) [inline python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py]: Can remove comment.

File: python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            print(f"{total_full_tiles_streamk=}")
+            print(f"{iters_per_tile=}")
+            print(f"{total_iters_streamk=}")
+            print("total_remainder_iters_streamk=", total_partial_tiles_streamk)
+        use_bias = False
+        # compute grid (work to do per SM on the first wave)
+        grids = total_programs_streamk
+        stride_bias = bias.stride(0) if use_bias else 0
+        #  P=P*0.0
+        #  locks=locks*0
```
Comment (neoblizz) [inline python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py]: We never need to reset locks, comment can be removed.

File: python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+#m, n, k = 1, 1024, 256
+#m, n, k = 8133, 8132, 8172  # some problem size to test
+#m, n, k = 8192, 8192, 8192  # some problem size to test
+#m, n, k = 8128, 6878, 7378  # some problem size to test
+#m, n, k = 8192, 4864, 6878  # some problem size to test
+#m, n, k = 512, 512, 512  # some problem size to test
+#m, n, k = 6912, 768, 256  # some problem size to test
+#m, n, k =4864, 8192, 4160  # some problem size to test
+#m, n, k = 5632, 6656, 7936
+m, n, k = 4864, 4096, 4300
```
Comment (neoblizz) [inline python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py:151]: Can we maybe create a list of "interesting m, n, k" and just have those listed here and why. For example, one of these is a golden m, n, k that divides perfectly onto the 304 CUs and the tile size.

**Review by neoblizz — CHANGES_REQUESTED** [review]

File: python/perf-kernels/streamk/streamk_kernel.py (hunk 2 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/streamk_kernel.py:166]: I will find a MI250 to test it.

File: python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py (hunk 7 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py:151]: done, see ea7b4050eea8532f3

File: python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py (hunk 6 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py]: removed

File: python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py (hunk 5 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/03-matrix-multiplication-stream-k.py]: done, thanks

File: python/perf-kernels/streamk/utils/file_generator.py (hunk 4 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/utils/file_generator.py:341]: can we leave this for the next release, as we need a thorough test for this, Thanks !

File: python/perf-kernels/streamk/streamk_kernel.py (hunk 3 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/streamk_kernel.py]: add NUM_XCDS tl.constexpr

File: python/perf-kernels/streamk/streamk_kernel.py (hunk 2 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/streamk_kernel.py:166]: add NUM_XCDS so we can switch on/off

File: python/perf-kernels/streamk/streamk_kernel.py (hunk 1 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/streamk_kernel.py:160]: not sure, how we are going to address this ?

File: python/perf-kernels/streamk/streamk_kernel.py (hunk 2 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/streamk_kernel.py:166]: sorry, NUM_XCDS can't help for cache_modifier

File: python/perf-kernels/streamk/streamk_kernel.py (hunk 1 above)
Comment (neoblizz) [reply python/perf-kernels/streamk/streamk_kernel.py:160]: I think there might be `if arch == gfx90a` which we can use for this and the pid renaming, I'll check.

File: python/perf-kernels/streamk/streamk_kernel.py (hunk 2 above)
Comment (neoblizz) [reply python/perf-kernels/streamk/streamk_kernel.py:166]: Yeah, we need something else. I'll investigate.

**Review by neoblizz — APPROVED** [review]

**Pull request merged**